lib/base.js

```diff
@@ -14,19 +15,33 @@ class BaseSchema {
         // Note: Joi will clone objects on changes and thus we need to update the schema reference
         let state = { schema: this._createSchema( engine ), engine };
 
-        for( let key in config ) {
+        // Check for when, and put it last as it changes type.
+        let keys = Object.keys(config);
+        if (keys.includes('when')) {
+          const index = keys.indexOf('when');
+          keys.splice(index, 1);
+          keys.push('when');
+        }
+
+        for (let i = 0; i < keys.length; i += 1) {
+            const key = keys[i];
 
-            this.updateSchema( state, key, config[ key ] );
+            this.updateSchema( state, key, config[ key ], engine );
         }
 
         return state.schema;
     }
 
-    updateSchema( state, key, value ) {
+    updateSchema( state, key, value, engine ) {
 
         if( utils.isFunction( state.schema[key] ) ) {
+            if( key === 'when' ) {
 
-            if( value === null ) {
+                const [ condition, options] = whenParser.parse(value, engine);
+
+                state.schema = state.schema[ key ]( condition, options );
+            }
+            else if( value === null ) {
```

Comment:
Since this was always sort of unclear, I'd recommend adding a comment here indicating what `state.schema[key]` is and why it is parsed uniquely in the `when` case.